README.md
# Rubber App Bar Flutter Package

## Description

The `rubber_app_bar` Flutter package is designed to enhance your app's user interface by providing a flexible and customizable Rubber App Bar. This package allows developers to create app bars that smoothly expand, contract, and transform based on user interactions.

## Key Features

- Smoothly expand, contract, and transform based on user interactions
- Customizable
- Flexible
- Easy to use

## Installation

```bash
  flutter pub add rubber_app_bar
```

## How to use

### 1. Put the `RubberAppBar` widget in the `appBar` property of the `Scaffold` widget

```dart
Scaffold(
  appBar: RubberAppBar(
    // ...
  ),
  // ...
)
```

### 2. Set the `height` property to the height of the app bar when it is collapsed

```dart
Scaffold(
  appBar: RubberAppBar(
    height: 200,
    // ...
  ),
  // ...
)
```

### 3. Set the `maxExtent` property to the height of the app bar when it is expanded

```dart
Scaffold(
  appBar: RubberAppBar(
    height: 200,
    maxExtent: 500,
    // ...
  ),
  // ...
)
```

### 4. Set the `builder` property to a function that returns the app bar's content (Your Custom Widget)

```dart
Scaffold(
  appBar: RubberAppBar(
    height: 200,
    maxExtent: 500,
    builder: (extending) => YourCustomWidget(),
    // ...
  ),
  // ...
)
```

### 5. Use the `extending` parameter to control the app bar's content

You will see that the builder gives you a parameter called `extending` which is the realtime expanding value of the app bar.
You can use this value to render your widgets conditionally based on the expanding value. For example, I want to render a widget with a height of 200 when the app bar is expanded to 200.

```dart
Scaffold(
  appBar: RubberAppBar(
    height: 200,
    maxExtent: 500,
    builder: (extending) => Column(
      children: [
        if (extending > 200)
          Container(
            height: 200,
            color: Colors.red,
          ),
      ],
    ),
    )
    // ...
  ),
```

### 5.A. Use the `extending` parameter in a better way

You can move, scale, rotate or do whatever you want with the widgets based on the `extending` value. For example, I want to scale up a widget from 0 to 1 as the app bar expands from 200 to 300.

```dart
Scaffold(
  appBar: RubberAppBar(
    height: 200,
    maxExtent: 500,
    builder: (extending) => Column(
      children: [
        if (extending > 200)
          Transform.scale(
            // (extending >= 300 ? 100 : (extending - 200)) == a range between [0, 100],
            // so we divide it by 100 to get [0, 1]
            scale: (extending >= 300 ? 100 : (extending - 200)) / 100,
            child: Container(
              height: 200,
              color: Colors.red,
            ),
          ),
      ],
    ),
    )
    // ...
  ),
```

### 6. Use the `mode` property to control the app bar's behavior

The default behavior of the app bar is to expand when the user drags it downwards and contract when the user drags it upwards. You can change this behavior by setting the `mode` property to `RubberAppBarMode.movementDirection` or `RubberAppBarMode.halfway`. The `RubberAppBarMode.movementDirection` mode will expand the app bar when the user drags it downwards and contract it when the user drags it upwards. The `RubberAppBarMode.halfway` mode will expand the app bar when the user drags it downwards and contract it when the user drags it upwards until it reaches the half of the `maxExtent` value, then it will expand the app bar when the user drags it upwards and contract it when the user drags it downwards.

```dart
Scaffold(
  appBar: RubberAppBar(
    height: 200,
    maxExtent: 500,
    mode: RubberAppBarMode.movementDirection,
    builder: (extending) => YourCustomWidget(),
    )
    // ...
  ),
```

### 7. Use the `transitionCurve` property to control the app bar's transition curve

The default transition curve of the app bar is `Curves.ease`. You can change this curve by setting the `transitionCurve` property to any curve you want.

```dart
Scaffold(
  appBar: RubberAppBar(
    height: 200,
    maxExtent: 500,
    transitionCurve: Curves.ease,
    builder: (extending) => YourCustomWidget(),
    )
    // ...
  ),
```

### 8. Use the `extendBodyBehindAppBar` property to control the app bar's position

The default position of the app bar is on top of the body. So as the app bar expands, it pushes the body down. You can change this behavior by setting the `extendBodyBehindAppBar` Scaffold's property to `true`. This will allow the app bar to be on top of the body. So as the app bar expands, it will be on top of the body.

```dart
Scaffold(
  extendBodyBehindAppBar: true,
  appBar: RubberAppBar(
    height: 200,
    maxExtent: 500,
    builder: (extending) => YourCustomWidget(),
    )
    // ...
  ),
```

### 9. Start your Scaffold body with a `SizedBox` widget with a height of `YourCustomWidget`'s height

In order to prevent the first top part of the body from being hidden behind the app bar, you need to start your Scaffold body with a `SizedBox` widget with a height of `YourCustomWidget`'s height. For example, if your `YourCustomWidget`'s height is 200, you need to start your Scaffold body with a `SizedBox` widget with a height of 200 as the first child in a `Column` for example.

```dart
Scaffold(
  extendBodyBehindAppBar: true,
  appBar: RubberAppBar(
    height: 200,
    maxExtent: 500,
    builder: (extending) => YourCustomWidget(),
    )
    body: Column(
      children: [
        // This is the first child
        SizedBox(height: 200),
        // The rest of the body
        // ...
      ],
    ),
  ),
```

## Example

```dart
import 'package:flutter/material.dart';
import 'package:rubber_app_bar/growing_widget.dart';
import 'package:rubber_app_bar/rubber_app_bar.dart';
import 'package:rubber_app_bar/scale_up_widget.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Travel App',
      home: HomeScreen(),
    );
  }
}

class HomeScreen extends StatelessWidget {
  const HomeScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: RubberAppBar(
        transitionCurve: Curves.ease,
        height: 200,
        maxExtent: 500,
        mode: RubberAppBarMode.movementDirection,
        builder: (extending) => Container(
          height: double.infinity,
          decoration: const BoxDecoration(
            color: Color.fromARGB(255, 9, 58, 255),
            borderRadius: BorderRadius.only(
              bottomLeft: Radius.circular(45),
              bottomRight: Radius.circular(45),
            ),
          ),
          child: SafeArea(
            child: Container(
              alignment: Alignment.center,
              padding: const EdgeInsets.all(16),
              child: Column(
                children: [
                  const Text(
                    'Rubber App Bar',
                    style: TextStyle(
                      fontSize: 30,
                      fontWeight: FontWeight.bold,
                      color: Color.fromARGB(255, 233, 233, 233),
                    ),
                  ),

                  // the condition is to render the widget only when
                  // the user is dragging the app bar downwards
                  // If you don't use this condition, you will get
                  // an overflow error when you try to drag the app bar upwards
                  if (extending > 0)
                    GrowingWidget(
                      availableHeight: extending,
                      maxHeight: 200,
                      child: const DummyWidget(
                        color: Color.fromARGB(255, 255, 204, 0),
                        height: 200,
                      ),
                    ),

                  // the condition is to render the widget only when
                  // the pervious widget is rendered and took its space
                  // 200 here represents the height of the previous widget
                  if (extending > 200)
                    // here we want to render two widgets,
                    // so we put them in a list and use the spread operator
                    ...[
                    const SizedBox(height: 20),
                    // extending - 200 is the available height for the widget
                    ScaleUpWidget(
                      availableHeight: extending - 200,
                      maxHeight: 150,
                      child: const DummyWidget(
                        color: Color.fromARGB(255, 255, 204, 0),
                        height: 150,
                      ),
                    )
                  ],
                  const Spacer(),
                  // Drag handle
                  const DragHandler(),
                ],
              ),
            ),
          ),
        ),
      ),

      // Here we are using this property to allow the app bar to be on top of the body
      // if you don't use it, the app bar will push the body down
      extendBodyBehindAppBar: true,
      body: const Center(
        child: Text(
          'This is the body!',
          style: TextStyle(
            fontSize: 30,
            fontWeight: FontWeight.bold,
            color: Color(0xFF2A2A2A),
          ),
        ),
      ),
    );
  }
}

class DragHandler extends StatelessWidget {
  const DragHandler({
    super.key,
  });

  @override
  Widget build(BuildContext context) {
    return Container(
      margin: const EdgeInsets.only(top: 16),
      width: 50,
      height: 5,
      decoration: BoxDecoration(
        color: const Color(0xFF2A2A2A),
        borderRadius: BorderRadius.circular(10),
      ),
    );
  }
}

class DummyWidget extends StatelessWidget {
  final Color color;
  final double height;
  const DummyWidget({
    super.key,
    required this.color,
    required this.height,
  });

  @override
  Widget build(BuildContext context) {
    return Container(
      height: height,
      decoration: BoxDecoration(
        color: color,
        borderRadius: BorderRadius.circular(10),
      ),
      child: const Center(
        child: Text(
          'Dummy Widget',
          style: TextStyle(
            fontSize: 22,
            fontWeight: FontWeight.bold,
            color: Color(0xFF2A2A2A),
          ),
        ),
      ),
    );
  }
}

```

## Demo

<img src="https://lh3.googleusercontent.com/drive-viewer/AEYmBYRtV34KNJV0a5gVmAuTdfi35EPQxbIDT7L0KBvuJ7VzvSbD476NxCAxPFL98kpCPjPfcXnzzIsr05qs5d_r1LgEXrmvhw=w2879-h1625" width="300" height="600" />


CHANGELOG.md
## 1.1.0

* Add `ScaleUpTransition` widget
* Fix initial `RubberAppBar` state bug

## 1.0.6

* General Enhancement and more documentation

## 1.0.3

* infinity height problem
* more documentation

## 1.0.2

* General Enhancement

## 1.0.1

* General Enhancement


test/rubber_app_bar_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:rubber_app_bar/rubber_app_bar.dart';

void main() {
  testWidgets('RubberAppBar expands and collapses correctly',
      (WidgetTester tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          appBar: RubberAppBar(
            height: 56.0,
            maxExtent: 200.0,
            builder: (double extent) {
              return Container(
                height: extent,
                color: Colors.blue,
              );
            },
          ),
          body: Container(),
        ),
      ),
    );

    // Verify initial state
// Verify initial state
    expect(find.byType(RubberAppBar), findsOneWidget);
    expect(find.byType(GestureDetector), findsOneWidget);
    expect(find.byType(AnimatedContainer), findsOneWidget);
    expect(find.byType(Container), findsNWidgets(3)); // Update this line

    expect(tester.getSize(find.byType(RubberAppBar).first).height,
        56.0); // Update this line

    // Expand the RubberAppBar
    await tester.drag(find.byType(GestureDetector), Offset(0, 300));
    await tester.pumpAndSettle();

    // Verify expanded state
    expect(tester.getSize(find.byType(Container).first).height,
        greaterThan(56.0)); // Update this line

    // Collapse the RubberAppBar
    await tester.drag(find.byType(GestureDetector), Offset(0, -300));
    await tester.pumpAndSettle();

    // Verify collapsed state
    expect(tester.getSize(find.byType(RubberAppBar).first).height,
        lessThanOrEqualTo(56.0)); // Update this line
  });
}


lib/rubber_app_bar.dart
/// A customizable app bar that can be expanded and collapsed with rubber-like behavior.
///
/// The [RubberAppBar] is a widget that provides a flexible app bar that can be expanded and collapsed
/// with a rubber-like behavior. It allows you to specify the initial height, maximum extent, and
/// transition duration of the app bar. You can also define callbacks for when the app bar is going up,
/// going down, reaching the top, or reaching the bottom.
///
/// The [RubberAppBar] can be used as a regular [AppBar] by setting the [mode] property to
/// [RubberAppBarMode.movementDirection]. In this mode, the app bar will expand and collapse based on
/// the direction of the user's gesture. Alternatively, you can set the [mode] property to
/// [RubberAppBarMode.halfWay] to make the app bar always expand halfway or collapse halfway.
///
/// To use the [RubberAppBar], simply wrap it around your content and provide a builder function that
/// returns the content based on the current extent of the app bar. The extent value can be used to
/// animate the content based on the app bar's height.
///
/// Example usage:
///
/// ```dart
/// RubberAppBar(
///   height: 100,
///   maxExtent: 200,
///   transitionDuration: Duration(milliseconds: 500),
///   mode: RubberAppBarMode.movementDirection,
///   builder: (double extent) {
///     return Container(
///       height: extent,
///       color: Colors.blue,
///       child: Text('Content'),
///     );
///   },
/// )
/// ```
import 'dart:math' show pow;
import 'package:flutter/material.dart';

/// Enum representing the mode of the RubberAppBar.
///
/// The [RubberAppBarMode] is the continuation mode after the user takes their finger off the screen.
/// - [halfWay]: if the app bar height is more than half of the maximum extent, the app bar will expand. Otherwise, it will collapse.
/// - [movementDirection]: if the user's gesture is moving up, the app bar will go up.
enum RubberAppBarMode {
  halfWay,
  movementDirection,
}

class RubberAppBar extends StatefulWidget implements PreferredSizeWidget {
  const RubberAppBar({
    Key? key,
    required this.height,
    required this.maxExtent,
    required this.builder,
    this.transitionDuration = const Duration(milliseconds: 500),
    this.transitionCurve = Curves.ease,
    this.onGoingUp,
    this.onGoingDown,
    this.onReachedTop,
    this.onReachedDown,
    this.mode = RubberAppBarMode.movementDirection,
  }) : super(key: key);

  final double height;
  final double maxExtent;
  final Duration transitionDuration;
  final Curve transitionCurve;
  final RubberAppBarMode mode;
  final void Function(double)? onGoingUp;
  final void Function(double)? onGoingDown;
  final VoidCallback? onReachedTop;
  final VoidCallback? onReachedDown;
  final Widget Function(double) builder;

  @override
  State<RubberAppBar> createState() => _RubberAppBarState();

  @override
  Size get preferredSize => Size.fromHeight(height + maxExtent);
}

class _RubberAppBarState extends State<RubberAppBar> {
  bool _isExpanded = false;
  double _currentExtent = 0;
  double dy = 0;

  void _updateExtent(double delta) {
    if (_currentExtent + dy > widget.maxExtent + 20) {
      _currentExtent = widget.maxExtent;
      return;
    } else if (_currentExtent + dy < -15) {
      _currentExtent = 0;
      return;
    }
    if (_currentExtent + dy > widget.maxExtent) {
      _currentExtent += dy / pow(_currentExtent, 0.5);
    } else if (_currentExtent + dy < 0) {
      _currentExtent += dy / pow(widget.maxExtent - _currentExtent, 0.5);
    } else {
      _currentExtent += dy;

      if (dy > 0) {
        widget.onGoingDown?.call((_currentExtent / widget.maxExtent) * 100);
      } else {
        widget.onGoingUp?.call((_currentExtent / widget.maxExtent) * 100);
      }
    }
  }

  void _updateExtentByDirection(double delta) {
    if (dy > 0) {
      _currentExtent = widget.maxExtent;

      Future.delayed(widget.transitionDuration, widget.onReachedDown);
    } else {
      _currentExtent = 0;
      Future.delayed(widget.transitionDuration, widget.onReachedTop);
    }
  }

  void _updateExtentHalfWay(double delta) {
    if (_currentExtent + dy > widget.maxExtent / 2) {
      _currentExtent = widget.maxExtent;
      Future.delayed(widget.transitionDuration, widget.onReachedDown);
    } else {
      _currentExtent = 0;
      Future.delayed(widget.transitionDuration, widget.onReachedTop);
    }
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onPanUpdate: (details) {
        setState(() {
          _isExpanded = true;
          dy = details.delta.dy;
          _updateExtent(dy);
        });
      },
      onPanEnd: (details) {
        setState(() {
          switch (widget.mode) {
            case RubberAppBarMode.halfWay:
              _updateExtentHalfWay(dy);
              break;
            case RubberAppBarMode.movementDirection:
              _updateExtentByDirection(dy);
              break;
          }
          _isExpanded = false;
        });
      },
      child: TweenAnimationBuilder<double>(
        duration: _isExpanded ? Duration.zero : widget.transitionDuration,
        curve: widget.transitionCurve,
        tween: Tween<double>(
          begin: _currentExtent + widget.height,
          end: _currentExtent + widget.height,
        ),
        builder: (BuildContext context, double value, Widget? child) {
          return SizedBox(
            height: value,
            child: widget.builder(value - widget.height),
          );
        },
      ),
    );
  }
}


lib/scale_up_widget.dart
import 'package:flutter/material.dart';

/// A widget that scales up its child based on the available height.
///
/// The [ScaleUpWidget] scales up its child widget based on the available height
/// and a maximum height value. It calculates the scale factor based on the
/// available height and applies it to the child widget using a [Transform.scale]
/// widget. The scaled child widget is then wrapped in a [SizedBox] to limit its
/// height and width based on the scale factor.
///
/// The [availableHeight] parameter specifies the available height for scaling.
/// The [child] parameter is the widget to be scaled up. The [maxHeight] parameter
/// defines the maximum height for the scaled child widget.
///
/// Example usage:
///
/// ```dart
/// ScaleUpWidget(
///   availableHeight: 300,
///   child: Container(
///     color: Colors.blue,
///   ),
///   maxHeight: 200,
/// )
/// ```
class ScaleUpWidget extends StatelessWidget {
  final double availableHeight;
  final Widget child;
  final double maxHeight;

  /// Creates a [ScaleUpWidget].
  ///
  /// The [availableHeight] parameter specifies the available height for scaling.
  /// The [child] parameter is the widget to be scaled up. The [maxHeight] parameter
  /// defines the maximum height for the scaled child widget.
  const ScaleUpWidget({
    super.key,
    required this.availableHeight,
    required this.child,
    required this.maxHeight,
  });

  @override
  Widget build(BuildContext context) {
    final double ratio = availableHeight / 100 - 1;
    final double scale = ratio >= 1 ? 1 : (ratio < 0 ? 0 : ratio);
    return LayoutBuilder(
      builder: (context, constraints) => Transform.scale(
        scale: scale,
        child: SizedBox(
          height: maxHeight * scale,
          width: constraints.maxWidth * scale,
          child: child,
        ),
      ),
    );
  }
}


lib/growing_container.dart
import 'package:flutter/material.dart';

/// A widget that grows to fit the available height up to a maximum height.
class GrowingWidget extends StatelessWidget {
  final double availableHeight;
  final double maxHeight;
  final Widget child;

  /// Creates a [GrowingWidget].
  ///
  /// The [child] is the widget to be displayed inside the [GrowingWidget].
  /// The [availableHeight] is the maximum height that the [GrowingWidget] can grow to.
  /// The [maxHeight] is the maximum height that the [GrowingWidget] can reach.
  const GrowingWidget({
    super.key,
    required this.child,
    required this.availableHeight,
    required this.maxHeight,
  });

  @override
  Widget build(BuildContext context) {
    final double height =
        availableHeight > maxHeight ? maxHeight : availableHeight;
    return SizedBox(
      height: height,
      child: child,
    );
  }
}


lib/scale_up_transistion.dart
import 'package:flutter/material.dart';

/// A widget that applies a scale-up animation to its child.
///
/// The animation starts from a scale of 0.0 and ends at a scale of 1.0.
/// The duration of the animation can be customized using the [duration] parameter.
/// The animation curve can be customized using the [curve] parameter.
/// An optional [delay] can be provided to delay the start of the animation.
///
/// Example usage:
/// ```dart
/// ScaleUp(
///   child: Container(
///     width: 100,
///     height: 100,
///     color: Colors.blue,
///   ),
///   duration: Duration(milliseconds: 500),
///   curve: Curves.easeInOut,
///   delay: Duration(milliseconds: 200),
/// )
/// ```
class ScaleUpTransition extends StatefulWidget {
  /// The child widget to apply the scale-up animation to.
  final Widget child;

  /// The duration of the animation.
  ///
  /// Defaults to 200 milliseconds.
  final Duration? delay;

  /// The delay before starting the animation.
  ///
  /// Defaults to no delay.
  final Duration duration;

  /// The curve of the animation.
  ///
  /// Defaults to [Curves.easeOutBack].
  final Curve curve;

  final double initialScale;
  final double endScale;

  const ScaleUpTransition({
    super.key,
    required this.child,
    this.delay,
    this.duration = const Duration(milliseconds: 200),
    this.curve = Curves.easeOutBack,
    this.initialScale = 0.0,
    this.endScale = 1.0,
  });

  @override
  State<ScaleUpTransition> createState() => _ScaleUpTransitionState();
}

class _ScaleUpTransitionState extends State<ScaleUpTransition>
    with TickerProviderStateMixin {
  late AnimationController _animationController;
  late Animation<double> _animation;

  @override
  void initState() {
    super.initState();
    _animationController = AnimationController(
      vsync: this,
      duration: widget.duration,
    );
    _animation =
        Tween<double>(begin: widget.initialScale, end: widget.endScale).animate(
      CurvedAnimation(
        parent: _animationController,
        curve: widget.curve,
      ),
    );

    Future.delayed(widget.delay ?? Duration.zero, () {
      _animationController.forward();
    });
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _animation,
      child: widget.child,
      builder: (context, child) {
        return Transform.scale(
          scale: _animation.value,
          child: child,
        );
      },
    );
  }
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:rubber_app_bar/growing_widget.dart';
import 'package:rubber_app_bar/rubber_app_bar.dart';
import 'package:rubber_app_bar/scale_up_widget.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Travel App',
      home: HomeScreen(),
    );
  }
}

class HomeScreen extends StatelessWidget {
  const HomeScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: RubberAppBar(
        transitionCurve: Curves.ease,
        height: 200,
        maxExtent: 500,
        mode: RubberAppBarMode.movementDirection,
        builder: (extending) => Container(
          height: double.infinity,
          decoration: const BoxDecoration(
            color: Color.fromARGB(255, 9, 58, 255),
            borderRadius: BorderRadius.only(
              bottomLeft: Radius.circular(45),
              bottomRight: Radius.circular(45),
            ),
          ),
          child: SafeArea(
            child: Container(
              alignment: Alignment.center,
              padding: const EdgeInsets.all(16),
              child: Column(
                children: [
                  const Text(
                    'Rubber App Bar',
                    style: TextStyle(
                      fontSize: 30,
                      fontWeight: FontWeight.bold,
                      color: Color.fromARGB(255, 233, 233, 233),
                    ),
                  ),

                  // the condition is to render the widget only when
                  // the user is dragging the app bar downwards
                  // If you don't use this condition, you will get
                  // an overflow error when you try to drag the app bar upwards
                  if (extending > 0)
                    GrowingWidget(
                      availableHeight: extending,
                      maxHeight: 200,
                      child: const DummyWidget(
                        color: Color.fromARGB(255, 255, 204, 0),
                        height: 200,
                      ),
                    ),

                  // the condition is to render the widget only when
                  // the pervious widget is rendered and took its space
                  // 200 here represents the height of the previous widget
                  if (extending > 200)
                    // here we want to render two widgets,
                    // so we put them in a list and use the spread operator
                    ...[
                    const SizedBox(height: 20),
                    // extending - 200 is the available height for the widget
                    ScaleUpWidget(
                      availableHeight: extending - 200,
                      maxHeight: 150,
                      child: const DummyWidget(
                        color: Color.fromARGB(255, 255, 204, 0),
                        height: 150,
                      ),
                    )
                  ],
                  const Spacer(),
                  // Drag handle
                  const DragHandler(),
                ],
              ),
            ),
          ),
        ),
      ),

      // Here we are using this property to allow the app bar to be on top of the body
      // if you don't use it, the app bar will push the body down
      extendBodyBehindAppBar: true,
      body: const Center(
        child: Text(
          'This is the body!',
          style: TextStyle(
            fontSize: 30,
            fontWeight: FontWeight.bold,
            color: Color(0xFF2A2A2A),
          ),
        ),
      ),
    );
  }
}

class DragHandler extends StatelessWidget {
  const DragHandler({
    super.key,
  });

  @override
  Widget build(BuildContext context) {
    return Container(
      margin: const EdgeInsets.only(top: 16),
      width: 50,
      height: 5,
      decoration: BoxDecoration(
        color: const Color(0xFF2A2A2A),
        borderRadius: BorderRadius.circular(10),
      ),
    );
  }
}

class DummyWidget extends StatelessWidget {
  final Color color;
  final double height;
  const DummyWidget({
    super.key,
    required this.color,
    required this.height,
  });

  @override
  Widget build(BuildContext context) {
    return Container(
      height: height,
      decoration: BoxDecoration(
        color: color,
        borderRadius: BorderRadius.circular(10),
      ),
      child: const Center(
        child: Text(
          'Dummy Widget',
          style: TextStyle(
            fontSize: 22,
            fontWeight: FontWeight.bold,
            color: Color(0xFF2A2A2A),
          ),
        ),
      ),
    );
  }
}
